### ISSUE TITLE
promote_type(Float16, Int128

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#50060]

**Comment by oscardssmith** [comment]
I think current behavior is right, but neither answer is great here.

**Comment by LilithHafner** [comment]
I think that `Float32` would be a reasonable choice here. `floatmax(Float32)` is about twice as high as `typemax(Int128)` and more than 99% as high as `typemax(UInt128)`.

**Comment by oscardssmith** [comment]
What do you think  `promote_type(Float32, BigInt)` should be?

**Comment by LilithHafner** [comment]
`Float64`'s 11 exponent bits are kinda sorta enough but `BigInt(floatmax(Float64))^2` is quite a reasonable `BigInt` to work with, so I'm not super happy with `Float64`. `Float32` would be terrible.

Something with 26 bits of mantissa and 37 bits of exponent might be ideal if it were an option, with a maximum normal value around 2^70,000,000,000.

Overall, `BigFloat` is a fine choice.

**Comment by elextr** [comment]
The documented rule is "Mixtures of integers and floating-point values are promoted to a floating-point type big enough to hold all the values.".  It would be surprising to users if this simple rule fails for some combinations of types.

So, if the integer is `Int128` the float must be `BigFloat`.

Of course thats not a performant choice, but if the user knows the range of values of the `Int128` is constrained they can specify the conversion manually.

**Comment by oscardssmith** [comment]
I don't think that rule is intended to mean without rounding (as 1+1.1 == BigFloat would completley suck). I think the intent is `typemax(TI<:Integer)<= floatmax(TF<:AbstractFloat)`.

**Comment by elextr** [comment]
The documentation definition is "Although it is not strictly necessary, it is generally implied that the common type to which the values are converted can faithfully represent all of the original values.  In this sense, the term "promotion" is appropriate since the values are converted to a "greater" type – i.e. one which can represent all of the input values in a single common type."

It would be strange if `Int128(66000) + Float16(1)` can't work correctly automatically, but it fails if the promotion type is `Float16`.

The real problem is `promote_type()` only has types to work on, not values so it has to assume maximums as pointed out above.  If actual values are considered the above could give `Float32(66001)` happily.  But Julia (and most other languages) don't work like that due to the ensuing type instability.

**Comment by stevengj** [comment]
Promoting to `Float32`, so that `typemax(Int128) ≤ floatmax(Float32)` seems reasonable to me.  That would also suggest promoting `Float16` with `Int32` and `Int64` to `Float32`.

A breaking change, however?

**Comment by gbaraldi** [comment]
Promoting to f32 does seem like an easy way to get type unstable code. A user using Float16 has to be aware that it's a very limited type and that mixing it with Int128 is probably a bad idea.

**Comment by stevengj** [comment]
Yes, people wouldn't expect
```jl
x = Float16(3.14159)
x += 1
```
to be type-unstable.  And if we don't promote to a larger precision for `Int64`, I'm not sure why we should do so for `Int128`.
